modules/library/render/src/main/java/org/geotools/renderer/lite/RendererUtilities.java

```diff
@@ -874,4 +876,105 @@ static FeatureCollection fixFeatureCollectionReferencing(FeatureCollection featu
         }
         return features;
     }
+
+    /**
+     * Finds a centroid for a polygon catching any exceptions resulting from generalization or 
+     * other polygon irregularities.
+     * 
+     * @param geom The polygon.
+     * @return The polygon centroid, or null if it can't be found.
+     */
+    public static Point getPolygonCentroid(Polygon geom) {
+        Point centroid;
+        try {
+            centroid = geom.getCentroid();
+        } catch (Exception e) {
+            // generalized polygons causes problems - this
+            // tries to hide them.
+            try {
+                centroid = geom.getExteriorRing().getCentroid();
+            } catch (Exception ee) {
+                try {
+                    centroid = geom.getFactory().createPoint(geom.getCoordinate());
+                } catch (Exception eee) {
+                    return null; // we're hooped
+                }
+            }
+        }
+        return centroid;
+    }
+
+    /**
+     * Uses a sampling technique to obtain a central point that lies inside the specified polygon.
+     * <p>
+     * Sampling occurs horizontally along the middle of the polygon obtained from the y coordinate of the polygon 
+     * centroid.
+     * </p>
+     * 
+     * @param geom The polygon.
+     * @param centroid The centroid of the polygon, can be null in which case it will be computed from {@link #getPolygonCentroid(Polygon)}.
+     * @param pg The prepared version of geom, can be null in which case it will be computed on demand.
+     * @param gf The geometry factory, can be null in which case the polygons factory will be used.
+     * 
+     * @return A central point that lies inside of the polygon, or null if one could not be found.
+     */
+    public static Point sampleForCentralPoint(Polygon geom, Point centroid, PreparedGeometry pg, GeometryFactory gf, 
```

Comment:
Both of those names make sense to me... I guess I'll go with sampleForInternalPoint, let me know if you like sampleForLabelPoint better and I'll happily rename.